### PR TITLE
drivers: gpio: Adding driver for gpio intel socfpga driver

### PR DESCRIPTION
**Pull request by ickochar**
- Added new gpio: gpio_intel_socfpga drivers for agilex socfpga
- Added gpio: gpio_intel_socfpga binding used in INTEL SOCFPGA AGILEX
- Adding gpio & pinmux node in INTEL SOCFPGA AGILEX dtsi file
- Added & updated codeowners for gpio driver

Signed-off-by: Chirag Kochar <chirag.kochar@intel.com>

### CODE REVIEW

File: drivers/gpio/gpio_intel_socfpga.c [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_intel_socfpga);
+
+/* Check if reset property is defined */
+#define GPIO_INTEL_INST_RESET_EN(inst)	DT_NODE_HAS_PROP(DT_DRV_INST(inst), resets)
+#define GPIO_INTEL_CHECK_RESET(inst)	GPIO_INTEL_INST_RESET_EN(inst) ||
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
+
+#if GPIO_INTEL_RESET_SUPPORT
```
Comment (cfriedt) [inline drivers/gpio/gpio_intel_socfpga.c:31]: Please remove the conditionals around the include

File: drivers/gpio/gpio_shell.c [hunk 2]
```diff
@@ -56,17 +58,33 @@ static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv)
...
-		index = (uint8_t)atoi(argv[args_indx.index]);
-		shell_print(sh, "Configuring %s pin %d",
+		cfg = (const struct gpio_driver_config *)dev->config;
+		if (cfg == NULL) {
+			shell_error(sh, "Device Configuration is NULL");
+			return -EIO;
+		}
+		if (index < find_msb_set(cfg->port_pin_mask)) {
+			shell_print(sh, "Configuring %s pin %u",
 			    argv[args_indx.port], index);
```
Comment (cfriedt) [inline drivers/gpio/gpio_shell.c:75]: Is this change necessary? It seems to be reaching into the implementation rather than using the interface. Shouldn't the new driver Just Work (TM) with `gpio_pin_configure()`

File: drivers/gpio/gpio_shell.c [hunk 3]
```diff
@@ -107,25 +138,43 @@ static int cmd_gpio_get(const struct shell *sh,
...
+		}
+		ret = gpio_pin_set(dev, index, value);
+		if (ret < 0) {
+			shell_error(sh, "Error %d while writing value", ret);
+			return -EIO;
+		}
+	} else {
+		shell_error(sh, "No such GPIO device");
+		return -ENODEV;
 	}
```
Comment (cfriedt) [inline drivers/gpio/gpio_shell.c:178]: Same here - I'm curious as to why you need to reach into the implementation here and not simply use `gpio_pin_set()`

File: drivers/gpio/gpio_shell.c [hunk 4]
```diff
@@ -138,24 +187,36 @@ static int cmd_gpio_set(const struct shell *sh,
...
+			return -EIO;
+		}
+		if (index < find_msb_set(cfg->port_pin_mask)) {
+			shell_fprintf(sh, SHELL_NORMAL,
+			"Blinking port %s pin %u.", argv[1], index);
+			shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
+		} else {
+			shell_error(sh, "Invalid GPIO pin");
+			return -EINVAL;
+		}
```
Comment (cfriedt) [inline drivers/gpio/gpio_shell.c:219]: Same here - it really seems like the changes to the GPIO shell are a workaround for something in the new driver.

**Review by cfriedt — CHANGES_REQUESTED** [review]
I think I would prefer to see this driver working without the changes to the GPIO shell.

File: drivers/gpio/gpio_shell.c (hunk 3 above)
Comment (ickochar) [reply drivers/gpio/gpio_shell.c:178]: The changes are needed if a zephyr application has enabled gpio shell. If a user passes a invalid pin number which is more then actual number of pins available,  then following results is observed currently :
agilex$ gpio get gpio@10c03200 24
Reading gpio@10c03200 pin 24
Value 0 
while valid pin values are from 0 to 23 only.

Fix is added in gpio_shell.c for 3 reasons :
1. Few API we don't pass pin values directly to the driver api. For example, gpio_pin_get will get whole port value and then do mask :
  int (*port_get_raw)(const struct device *port,
			      gpio_port_value_t *value);

	  int (*port_set_masked_raw)(const struct device *port,
				     gpio_port_pins_t mask,
				     gpio_port_value_t value);
2. All the driver might need to have same functionality when shell is implemented.
3. Values passed above 32, will always give value 0 even if currently no. of pins above 32 doesn't exits.

New driver doesn't this this change for valid pins. This is just for invalid pins.

File: drivers/gpio/gpio_shell.c (hunk 2 above)
Comment (ickochar) [reply drivers/gpio/gpio_shell.c:75]: same explanation for other comment

File: drivers/gpio/gpio_shell.c (hunk 4 above)
Comment (ickochar) [reply drivers/gpio/gpio_shell.c:219]: same explanation for other comment

**Comment by ickochar** [comment]
> I think I would prefer to see this driver working without the changes to the GPIO shell.

@cfriedt Yes the driver works fine independent to shell chances.

File: drivers/gpio/gpio_intel_socfpga.c [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_intel_socfpga);
+
+/* Check if reset property is defined */
+#define GPIO_INTEL_INST_RESET_EN(inst)	DT_NODE_HAS_PROP(DT_DRV_INST(inst), resets)
+#define GPIO_INTEL_CHECK_RESET(inst)	GPIO_INTEL_INST_RESET_EN(inst) ||
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
```
Comment (cfriedt) [inline drivers/gpio/gpio_intel_socfpga.c:29]: Please place instance macros near the bottom (a convention in Zephyr) along with other instance macros

File: drivers/gpio/gpio_intel_socfpga.c (hunk 1 above)
Comment (cfriedt) [reply drivers/gpio/gpio_intel_socfpga.c:31]: I gathered that. Can you please remove the conditionals around the include? It's something we try to avoid.

**Review by cfriedt — CHANGES_REQUESTED** [review]
Please make changes to the gpio shell in a separate PR and make the remaining minor changes requested.

**Comment by ickochar** [comment]
Could we re-open this review ?

File: drivers/gpio/gpio_intel_socfpga.c [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+struct gpio_intel_socfpga_config {
+	/** GPIO common config */
+	struct gpio_driver_config gpio_config;
+	/** GPIO MMIO mapped physical address */
+	DEVICE_MMIO_NAMED_ROM(gpio_mmio);
+	/** GPIO port number (0 or 1) */
+	const uint32_t gpio_port;
+	/** GPIO pin number per port */
+	uint32_t ngpios;
+#if WDT_DW_RESET_SUPPORT
```
Comment (cfriedt) [inline drivers/gpio/gpio_intel_socfpga.c:59]: This seems out of place. Is there a Kconfig value missing? I can't see this symbol being used anywhere in Zephyr currently.

File: soc/arm64/intel_socfpga/agilex/mmu_regions.c [hunk 7]
```diff
@@ -26,6 +26,12 @@ static const struct arm_mmu_region mmu_regions[] = {
 			      DT_REG_SIZE(DT_NODELABEL(uart0)),
 			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE),
 
+	MMU_REGION_FLAT_ENTRY("PINMUX",
```
Comment (cfriedt) [inline soc/arm64/intel_socfpga/agilex/mmu_regions.c:29]: I'm not super familiar with this subject, so perhaps @gmarull could take a look.

File: drivers/gpio/gpio_intel_socfpga.c (hunk 1 above)
Comment (cfriedt) [reply drivers/gpio/gpio_intel_socfpga.c:31]: ^^ please make the suggested changes

File: drivers/gpio/gpio_intel_socfpga.c (hunk 5 above)
Comment (cfriedt) [reply drivers/gpio/gpio_intel_socfpga.c:29]: ^^ Please address previous review comments.

Part of the reason that DT instance macros are typically defined at the bottom of the file are so that it's easier to determine if they are actually being used.

It seems to me that the above two macros are unused. They should be removed if that's the case.

**Review by cfriedt — CHANGES_REQUESTED** [review]
I would have personally left this PR closed as it grew stale since the author ignored previous reviews, had not made requested changes, and accrued merge conflicts.

Since it has been reopened though.. still the same thing. Please address review comments.

I did find additional items of concern as well, so I highlighted those too.

**Comment by ickochar** [comment]
> I think I would prefer to see this driver working without the changes to the GPIO shell.
Separated to new PR

File: drivers/gpio/gpio_intel_socfpga.c (hunk 6 above)
Comment (ickochar) [reply drivers/gpio/gpio_intel_socfpga.c]: Updated to correct GPIO based Macro

**Comment by ickochar** [comment]
> I would have personally left this PR closed as it grew stale since the author ignored previous reviews, had not made requested changes, and accrued merge conflicts.
Apologies for inconvenience. I have addressed all the review comments all other in last push.

File: drivers/gpio/gpio_intel_socfpga.c (hunk 5 above)
Comment (ickochar) [reply drivers/gpio/gpio_intel_socfpga.c:30]: As the GPIO_INTEL_RESET_SUPPORT macro is used in other part of the driver file, I have created this macro at the beginning of the file.

**Comment by ickochar** [comment]
Hello All,

Could you please help to review this PR @cfriedt @carlocaione @gmarull @SgrrZhf

**Comment by ickochar** [comment]
I would like to request to add dev-review label.

**Review by cfriedt — APPROVED** [review]
No particular red flags for me at this point. Sorry I have not been active as GPIO collaborator for some time.

File: drivers/gpio/gpio_intel_socfpga.c [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_intel_socfpga);
+
+/* Check if reset property is defined */
+#define GPIO_INTEL_INST_RESET_EN(inst)	DT_NODE_HAS_PROP(DT_DRV_INST(inst), resets)
+#define GPIO_INTEL_CHECK_RESET(inst)	GPIO_INTEL_INST_RESET_EN(inst) ||
+#define GPIO_INTEL_RESET_SUPPORT	(DT_INST_FOREACH_STATUS_OKAY(GPIO_INTEL_CHECK_RESET) 0)
+
+#define DEV_CFG(_dev) ((const struct gpio_intel_socfpga_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct gpio_intel_socfpga_data *const)(_dev)->data)
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_intel_socfpga.c:33]: Please avoid macros like this. Just access the structure members directly.

File: drivers/gpio/gpio_intel_socfpga.c [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+		 * pin 40 - 0x10D1 3100
+		 * ...
+		 * Below GAP is added after pin 40.
+		 */
+		if (hps_io_pin >= PMUX_40_SEL) {
+			addroffst += PMUX_40_SEL_GAP;
+		}
+
+		/* Getting pinmux base address from pinmux node */
+		pmux_value = sys_read32((DT_REG_ADDR(DT_NODELABEL(pinmux)) + addroffst));
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_intel_socfpga.c:130]: I'd avoid using a these macros deep in the code as it hides the dependency.

File: drivers/gpio/gpio_intel_socfpga.c [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+	sys_write32(value, reg_base);
+
+	return 0;
+}
+
+static int gpio_socfpga_port_set_masked_raw(const struct device *port,
+					 gpio_port_pins_t mask,
+					 gpio_port_value_t value)
+{
+	return -ENOSYS;
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_intel_socfpga.c:238]: Why?

File: drivers/gpio/gpio_intel_socfpga.c [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+
+#define CREATE_GPIO_DEVICE(_inst)							\
+	static struct gpio_intel_socfpga_data gpio_data_##_inst;			\
+	static const struct gpio_intel_socfpga_config gpio_config_##_inst = {		\
+		DEVICE_MMIO_NAMED_ROM_INIT(gpio_mmio, DT_DRV_INST(_inst)),		\
+		.gpio_config.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(_inst),	\
+		.gpio_port = _inst,							\
+		.ngpios = DT_INST_PROP(_inst, ngpios),					\
+		IF_ENABLED(DT_INST_NODE_HAS_PROP(inst, resets),				\
+			(DW_WDT_RESET_SPEC_INIT(inst)))					\
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_intel_socfpga.c:294]: WDT?

**Review by nashif — APPROVED** [review]

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
There seem to be some confusion regarding the added pinmux. Should it be using the pinctrl driver API? Should it be using the syscon API? Current access to it is done directly, which hides the dependency deep with the code.

File: drivers/gpio/gpio_intel_socfpga.c (hunk 5 above)
Comment (MaureenHelm) [reply drivers/gpio/gpio_intel_socfpga.c:30]: I agree with @cfriedt . You shouldn't be using DT inst macros within the `struct gpio_intel_socfpga_config` or `gpio_init` definitions.

File: samples/basic/blinky/boards/intel_socfpga_agilex_socdk.overlay [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+		compatible = "gpio-leds";
+		myled0: led_0 {
+			gpios = <&gpio1 19 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&gpio1 {
+	status = "okay";
+};
```
Comment (MaureenHelm) [inline samples/basic/blinky/boards/intel_socfpga_agilex_socdk.overlay:24]: Why not put these in the board devicetree?

File: CODEOWNERS [hunk 13]
```diff
@@ -234,6 +234,7 @@
 /drivers/flash/*nrf*                      @de-nordic
 /drivers/flash/*esp32*                    @sylvioalves
 /drivers/gpio/*b91*                       @andy-liu-telink
+/drivers/gpio/*intel_socfpga*             @alifzakuan @ickochar
```
Comment (MaureenHelm) [inline CODEOWNERS:237]: Use the MAINTAINERS file instead of CODEOWNERS

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_intel_socfpga.c (hunk 5 above)
Comment (ickochar) [reply drivers/gpio/gpio_intel_socfpga.c:30]: @MaureenHelm 
Same driver has to support multiple platform (i.e. agilex7 & agilex5) which is the main factor for implementation like this. Its possible that platform could have reset manager and some platform could not have reset manager support. So we have to check if the reset manager property is defined in a node. If yes, then we should toggle the reset manager pin to reset the peripheral during initialization of the driver.

File: drivers/gpio/gpio_intel_socfpga.c (hunk 8 above)
Comment (ickochar) [reply drivers/gpio/gpio_intel_socfpga.c:33]: We are memory mapping device register using DEVICE_MMIO_NAMED_MAP macros and this macro requires DEV_DATA and DEV_CFG macros to be defined. Please check below link :

https://github.com/zephyrproject-rtos/zephyr/blob/71fd9b2e07a6fa9d7895759fc83685b6849aebb4/include/zephyr/sys/device_mmio.h#L520

File: drivers/gpio/gpio_intel_socfpga.c (hunk 9 above)
Comment (ickochar) [reply drivers/gpio/gpio_intel_socfpga.c:130]: In understand your main concern related to complete pinmux/pinctrl implementation. 
Currently another hardware outside zephyr world, configuring pin controller register on our platform. 
Our main design goal is to make sure pinctrl is configured in GPIO mode before performing any GPIO operation. 
Also, Unfortunately, Pinmux driver does not have get pinctrl register values. If you say, I can create a RFC for pinctrl enhancement separately. Is that fine ?

File: samples/basic/blinky/boards/intel_socfpga_agilex_socdk.overlay (hunk 12 above)
Comment (ickochar) [reply samples/basic/blinky/boards/intel_socfpga_agilex_socdk.overlay:24]: We are currently try to enable as minimum as possible in board dtsi file. GPIO is not an essential peripheral for our platform bring up. So by default all non-essential peripheral are turned off. 
Also our pins could be re-configured for various other purpose like SPI, I2C, SDMMC , etc, Depending on end application use case, respective peripherals could be enabled flexibly. For example as this is blinky application, we can enable only GPIO peripheral here with required alias for sample application.
